docs/src/tutorial-nondeterministic-variables.md

```diff
@@ -1,95 +1,114 @@
 
 ```rust
 {{#include tutorial/arbitrary-variables/src/inventory.rs:inventory_lib}}
 ```
 
-Now we would like to verify that, no matter which combination of `id` and `quantity`, a call to `Inventory::update()` followed by a call to `Inventory::get()` using the same `id` returns some value that's equal to the one we inserted:
+Let's write a fairly simple proof harness, one that just ensures we successfully `get` the value we inserted with `update`:
 
 ```rust
 {{#include tutorial/arbitrary-variables/src/inventory.rs:safe_update}}
 ```
 
-In this harness, we use `kani::any()` to generate the new `id` and `quantity`.
-`kani::any()` is a **safe** API function, and it represents only valid values.
+We use `kani::any()` twice here:
 
-If we run this example, Kani verification will succeed, including the assertion that shows that the underlying `u32` variable  used to represent `NonZeroU32` cannot be zero, per its type invariant:
+1. `id` has type `ProductId` which was actually just a `u32`, and so any value is fine.
+2. `quantity`, however, has type `NonZeroU32`.
+In Rust, it would be undefined behavior to have a value of `0` for this type.
 
-You can try it out by running the example under
-[`arbitrary-variables`](https://github.com/model-checking/kani/tree/main/docs/src/tutorial/arbitrary-variables/):
+We included an extra assertion that the value returned by `kani::any()` here was actually non-zero.
+If we run this, you'll notice that verification succeeds.
 
-```
+```bash
 cargo kani --harness safe_update
 ```
 
-## Unsafe nondeterministic variables
+`kani::any()` is safe Rust, and so Kani only implements it for types where type invariants are enforced.
+For `NonZeroU32`, this means we never return a `0` value.
+The assertion we wrote in this harness was just an extra check we added to demonstrate this fact.
 
-Kani also includes an **unsafe** method to generate unconstrained nondeterministic variables which do not take type invariants into consideration.
-As with any unsafe method in Rust, users have to make sure the right guardrails are put in place to avoid undesirable behavior.
+## Bounding nondeterministic variables
 
-That said, there may be cases where you want to verify your code taking into consideration that some inputs may contain invalid data.
+There is no implementation of `kani::any()` for more complex data structures like `Vec`, however.
+The trouble with a nondeterministic vector is that you usually need to _bound_ the size of the vector, for the reasons we investigated in the [last chapter](./tutorial-loop-unwinding.md).
+The `kani::any()` function does not have any arguments, and so cannot be given an upper bound.
 
-Let's see what happens if we modify our verification harness to use the unsafe method `kani::any_raw()` to generate the updated value.
+You can use `kani::any()` for `[T; N]` (if implemented for `T`).
```

Comment:
This line doesn't make much sense here.

What if you move talk about custom non-deterministic types first then you introduce this section. I think it is easier to first present the solution and discuss its limitation after.